Python ij review (#10)
ome/omero-guide-fiji: docs/imagej_python.rst

Using the Python API allows us to easily load the 2D-plane we need to see or analyze.

> This is much easier that when using the Java API and Bio-Formats plugin.

Reviewer: ```suggestion
This is much easier than using the Java API and Bio-Formats plugin.
```

We will show in the examples:

- How to start ImageJ in Python.

- How to load data from OMERO.

- How to run ImageJ macro from Python.

Setup
-----

For using the Python examples and notebooks of this guide we recommend using Conda (Option 1).
Conda manages programming environments in a manner similar to 
`virtualenv <https://virtualenv.pypa.io/en/stable/>`_.

Alternatively you can use ``repo2docker`` to build and run a Docker image locally (Option 2).
This Docker image will provide the Conda environment and Jupyter notebooks with some image 
analysis workflows.

*Option 1*
~~~~~~~~~~

Install omero-py and `pyimagej <https://pypi.org/project/pyimagej/>`_ via Conda:

- Install `Miniconda <https://docs.conda.io/en/latest/miniconda.html>`_ if necessary.

- Create a programming environment using Conda and activate it::

    $ conda create -n imagej_python python=3.6

    $ conda activate imagej_python

- Install omero-py and `pyimagej <https://pypi.org/project/pyimagej/>`_::

    $ conda install -c conda-forge pyimagej

    $ conda install -c ome omero-py

> Note: we have noticed few problems using that approach when running the scripts on some operating systems due to some issues with the ImageJ1-ImageJ2 bridge.

Reviewer: ```suggestion
Note: we have noticed problems using the Option 1 approach when running the scripts on some operating systems due to issues with the ImageJ1-ImageJ2 bridge.
```

*Option 2*
~~~~~~~~~~

Create a local Docker Image using ``repo2docker``, see :download:`README.md <https://github.com/ome/omero-guide-fiji/blob/master/README.md>`::

    $ pip install jupyter-repo2docker
    $ git clone https://github.com/ome/omero-guide-fiji.git
    $ cd omero-guide-fiji
    $ repo2docker .

When the Image is ready:

- Copy the URL displayed in the terminal in your favorite browser

- Click the ``New`` button on the right-hand side of the window

- Select ``Terminal`` 

    \ |image0|

- A Terminal will open in a new Tab

- A Conda environment has already been created when the Docker Image was built

- To list all the Conda environments, run::

    $ conda env list

- The environment with the OMERO Python bindings and a few other libraries is named ``notebook``, activate it::

    $ conda activate notebook

Step-by-Step
------------

The script used in this document is :download:`run_macro_python.py <../scripts/python/run_macro_python.py>`.

> One of the advantage of this approach is that we can load only the 2D-planes we wish to analyze.

Reviewer: ```suggestion
One of the advantages of this approach is that we can load only the 2D-planes we wish to analyze.
```

The script used in this document contains an ImageJ1 macro that needs graphical user interface (GUI) elements, and thus it requires using ImageJ in GUI mode. In this GUI mode, the resulting windows content is handled.
If you are running the example in the Docker container, 
you will also need to start UI environment if it is not already up.

> If you have used the option 2 above, select ``desktop`` to start the UI environment.

Reviewer: ```suggestion
If you have used the Option 2 above, select ``desktop`` to start the UI environment.
```

    \ |image0|

> If you are using the option 1 above, you will first need to update the script to point to your local installation of Fiji or use one of the options described in `ImageJ Tutorials <https://nbviewer.jupyter.org/github/imagej/tutorials/blob/master/notebooks/ImageJ-Tutorials-and-Demo.ipynb>`_.

Reviewer: ```suggestion
If you are using the Option 1 above, you will first need to update the script to point to your local installation of Fiji or use one of the options described in `ImageJ Tutorials <https://nbviewer.jupyter.org/github/imagej/tutorials/blob/master/notebooks/ImageJ-Tutorials-and-Demo.ipynb>`_.
```

We will show in the notebooks:

- How to start ImageJ in Python.

- How to load data from OMERO.

- How to run ImageJ macro from Python.

Setup
-----

For using the Python examples and notebooks of this guide we recommend using Conda (Option 1).
Conda manages programming environments in a manner similar to 
`virtualenv <https://virtualenv.pypa.io/en/stable/>`_.

Alternatively you can use ``repo2docker`` to build and run a Docker image locally (Option 2).
This Docker image will provide the Conda environment and Jupyter notebooks with some image 
analysis workflows.

*Option 1*
~~~~~~~~~~

Install omero-py and `pyimagej <https://pypi.org/project/pyimagej/>`_ via Conda:

- Install `Miniconda <https://docs.conda.io/en/latest/miniconda.html>`_ if necessary.

- Create a programming environment using Conda and activate it::

    $ conda create -n imagej_python python=3.6

    $ conda activate imagej_python

- Install omero-py and `pyimagej <https://pypi.org/project/pyimagej/>`_::

    $ conda install -c conda-forge pyimagej

    $ conda install -c ome omero-py

*Option 2*
~~~~~~~~~~

> Create a local Docker Image using ``repo2docker``, see :download:`README.md <https://github.com/ome/omero-guide-python/blob/master/README.md>`::

Reviewer: In the README, which is hosted in omero-guides-python is a sentence "clone this repository". But in the sense of this workflow, the omero-guides-python repo does not need to be cloned (note that when you click onto the README, it does not download the file locally, instead, opens it in github)

    $ pip install jupyter-repo2docker
    $ git clone https://github.com/ome/omero-guide-fiji.git
    $ cd omero-guide-fiji
    $ repo2docker .

When the Image is ready:

- Copy the URL displayed in the terminal in your favorite browser

- Click the ``New`` button on the right-hand side of the window

- Select ``Terminal``

.. image:: images/terminal.png

- A Terminal will open in a new Tab

- A Conda environment has already been created when the Docker Image was built

- To list all the Conda environments, run::

    $ conda env list

- The environment with the OMERO Python bindings and a few other libraries is named ``notebook``, activate it::

    $ conda activate notebook

Step-by-Step
------------

The script used in this document is :download:`run_macro_python.py <../scripts/python/run_macro_python.py>`.
One of the advantage of this approach is that we can load only the 2D-planes we wish to analyze.

In this example, Fiji has been installed locally. It is possible to install Fiji "on the fly",
see `ImageJ Tutorials <https://nbviewer.jupyter.org/github/imagej/tutorials/blob/master/notebooks/ImageJ-Tutorials-and-Demo.ipynb>`_ for more options.

> To run the macro, we use ImageJ1 windows so it requires using ImageJ in GUI mode and requires handling the resulting windows. This is the reason the parameter `headless` is set to `False`.

Reviewer: ```suggestion
To run the macro, we use ImageJ1 graphical user interface (GUI) elements, and thus it requires using ImageJ in GUI mode. In this GUI mode, the resulting windows content is handled. This is the reason for the parameter `headless` to be set to `False`.
```

Reviewer: I think also this paragraph needs to be nearer to the code block below where you are setting the headless parameter.

Author: done

The script used in this document contains an ImageJ1 macro that needs graphical user interface (GUI) elements, and thus it requires using ImageJ in GUI mode. In this GUI mode, the resulting windows content is handled.
If you are running the example in the Docker container, 
you will also need to start UI environment if it is not already up.
If you have used the option 2 above, select ``desktop`` to start the UI environment.

    \ |image0|

> If you do not use any ImageJ1 features e.g. macro, you do **not** need the UI environment and do **not** need to set the `headless` parameter (default is `True`).

Reviewer: ```suggestion

```

Reviewer: This sentence - why do we need that ? It still creates confusion.

Author: If someone is not interested to run the macro then this is not needed.
I could write a second script and point to it.

    \ |image0|

> If you do not use any ImageJ1 features e.g. macro, you do **not** need the UI environment.

Reviewer: What if you put that sentence below the "You can now run the script."
?

For that script you need the UI environment for sure.
